hedera-node/src/main/java/com/hedera/services/contracts/sources/TxnAwareEvmSigsVerifier.java

```diff
@@ -26,126 +26,115 @@
 	private final TransactionContext txnCtx;
 	private final BiPredicate<JKey, TransactionSignature> cryptoValidity;
-	private final Supplier<MerkleMap<EntityNum, MerkleToken>> tokens;
-	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
 
 	@Inject
-	public TxnAwareSoliditySigsVerifier(
+	public TxnAwareEvmSigsVerifier(
 			final ActivationTest activationTest,
 			final TransactionContext txnCtx,
-			final Supplier<MerkleMap<EntityNum, MerkleToken>> tokens,
-			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
 			final BiPredicate<JKey, TransactionSignature> cryptoValidity
 	) {
 		this.txnCtx = txnCtx;
-		this.tokens = tokens;
-		this.accounts = accounts;
 		this.activationTest = activationTest;
 		this.cryptoValidity = cryptoValidity;
 	}
 
 	@Override
 	public boolean hasActiveKey(
-			final Address accountAddress,
-			final Address recipient,
-			final Address contract,
-			final Address activeContract,
-			final ContractAliases aliases
+			final boolean isDelegateCall,
+			@NotNull final Address accountAddress,
+			@NotNull final Address activeContract,
+			@NotNull final WorldLedgers worldLedgers
 	) {
 		final var accountId = EntityIdUtils.accountIdFromEvmAddress(accountAddress);
-		final var simpleId = Id.fromGrpcAccount(accountId);
-		final var entityNum = simpleId.asEntityNum();
-		final var account = accounts.get().get(entityNum);
-		validateTrue(account != null, INVALID_ACCOUNT_ID);
+		final var account = Optional.ofNullable(worldLedgers.accounts().getImmutableRef(accountId));
+
+		validateTrue(account.isPresent(), INVALID_ACCOUNT_ID);
 
 		if (accountAddress.equals(activeContract)) {
 			return true;
 		}
 
-		return isActiveInFrame(account.getAccountKey(), recipient, contract, activeContract, aliases);
+		final MerkleAccount merkleAccount = account.map(MerkleAccount.class::cast).orElseGet(MerkleAccount::new);
+		return merkleAccount.getAccountKey() != null && isActiveInFrame(merkleAccount.getAccountKey(), isDelegateCall,
+				activeContract,
+				worldLedgers.aliases());
 	}
 
 	@Override
 	public boolean hasActiveSupplyKey(
-			final Address tokenAddress,
-			final Address recipient,
-			final Address contract,
-			final Address activeContract,
-			final ContractAliases aliases
+			final boolean isDelegateCall,
+			@NotNull final Address tokenAddress,
+			@NotNull final Address activeContract,
+			@NotNull final WorldLedgers worldLedgers
 	) {
 		final var tokenId = EntityIdUtils.tokenIdFromEvmAddress(tokenAddress);
-		final var simpleId = Id.fromGrpcToken(tokenId);
-		final var entityNum = simpleId.asEntityNum();
-		final var token = tokens.get().get(entityNum);
-		validateTrue(token != null, INVALID_TOKEN_ID);
-		validateTrue(token.hasSupplyKey(), TOKEN_HAS_NO_SUPPLY_KEY);
-		return isActiveInFrame(token.getSupplyKey(), recipient, contract, activeContract, aliases);
+		final var token =
+				worldLedgers.tokens() != null ? Optional.ofNullable(worldLedgers.tokens().getImmutableRef(tokenId)) :
```

Comment:
Note we only need two pieces of information here:
1. Does the token exist?
2. What is the token's `supplyKey` (if any)?

So using `tokens.getImmutableRef(tokenId)` is actually quite wasteful. Depending on the depth of the `MessageFrame` stack, it creates **at least two** new `MerkleToken`s; and for every property [enumerated by `TokenProperty`](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/ledger/properties/TokenProperty.java#L34), calls `getter()`/`setter()` pairs to fully initialize these `MerkleToken`s.

Instead we can directly get (1) by calling `tokens.exists()` and (2) by calling `tokens.get()`. E.g.,
```
validateTrue(tokens.exists(tokenId), INVALID_TOKEN_ID);
final var supplyKey = (JKey) tokens.get(tokenId, TokenProperty.SUPPLY_KEY);
validateTrue(supplyKey != null, TOKEN_HAS_NO_SUPPLY_KEY);

return isActiveInFrame(supplyKey, isDelegateCall, activeContract, worldLedgers.aliases());
```

Reply:
Thank you for noticing this performance issue! I applied the fix you suggested, not only for the tokens's SUPPLY_KEY, but also for the account's KEY.